### PR TITLE
Add wrapper around Linux perf events. Renamed CPU timer to reflect rusage

### PR DESCRIPTION
**Pull request by mbutrovich**
Now have rusage_monitor and perf_monitor giving us access to perf_events.

### CODE REVIEW

File: src/include/common/perf_monitor.h [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+   * Represents the struct read_format with PERF_FORMAT_GROUP enabled, PERF_FORMAT_TOTAL_TIME_ENABLED and
+   * PERF_FORMAT_TOTAL_TIME_RUNNING disabled. http://www.man7.org/linux/man-pages/man2/perf_event_open.2.html
+   */
+  struct PerfCounters {
+    /**
+     * Number of events read. This should be PerfMonitor::NUM_HW_EVENTS if working, 0 otherwise
+     */
+    uint64_t num_events_;
+    /**
+     * Total cycles. Be wary of what happens during CPU frequency scaling.
```
Comment (malin1993ml) [inline src/include/common/perf_monitor.h]: What does this mean? Can you be a little bit more specific?

File: src/include/common/perf_monitor.h [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    // https://lkml.org/lkml/2018/2/13/810
+    // uint64_t branch_instructions_;
+    // uint64_t branch_misses_;
+    /**
+     * Bus cycles, which can be different from total cycles.
+     */
+    uint64_t bus_cycles_;
+    /**
+     * Total cycles; not affected by CPU frequency scaling.
+     */
```
Comment (malin1993ml) [inline src/include/common/perf_monitor.h]: What do you mean by "not affected?"

File: src/include/common/perf_monitor.h [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+#else
+    // Initialize perf configuration
+    perf_event_attr pe;
+    std::memset(&pe, 0, sizeof(perf_event_attr));
+    pe.type = PERF_TYPE_HARDWARE;
+    pe.size = sizeof(perf_event_attr);
+    pe.disabled = 1;
+    pe.exclude_kernel = 1;
+    pe.exclude_hv = 1;
+    pe.read_format = PERF_FORMAT_GROUP;
```
Comment (malin1993ml) [inline src/include/common/perf_monitor.h]: Can you explain what are these?

File: src/include/common/perf_monitor.h [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+    // Open file descriptors for each perf_event that we want. We reuse the first entry of the array as the group fd.
+    for (uint8_t i = 0; i < NUM_HW_EVENTS; i++) {
+      pe.config = HW_EVENTS[i];
+      event_files_[i] = syscall(__NR_perf_event_open, &pe, 0, -1, event_files_[0], 0);
```
Comment (malin1993ml) [inline src/include/common/perf_monitor.h]: And what is this doing?

File: src/include/common/perf_monitor.h (hunk 1 above)
Comment (mbutrovich) [reply src/include/common/perf_monitor.h]: Just copying from the man page.

File: src/include/common/perf_monitor.h (hunk 2 above)
Comment (mbutrovich) [reply src/include/common/perf_monitor.h]: Just copying from the man page.

File: src/include/common/perf_monitor.h (hunk 3 above)
Comment (mbutrovich) [reply src/include/common/perf_monitor.h]: I would generally just refer people to the man page that I link at the top of the file. I'm not sure how useful it is to reproduce a ton of the man page. I did it where absolutely necessary.

File: src/include/common/perf_monitor.h (hunk 4 above)
Comment (mbutrovich) [reply src/include/common/perf_monitor.h]: It's opening file descriptors for each perf_event. See line 111.

File: test/common/perf_monitor_test.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  }
+};
+
+/**
+ * Simple test that spins off 2 threads. One thread builds and then destroys a Catalog. The other sleeps. The parent
+ * waits. We then do comparisons on their respective perf counters based on the work performed.k
+ */
+// NOLINTNEXTLINE
+TEST_F(PerfMonitorTests, BasicTest) {
+  common::PerfMonitor monitor(true);
```
Comment (malin1993ml) [inline test/common/perf_monitor_test.cpp]: Should we have a separate test to test `monitor(false)`? Would that work here?

**Comment by malin1993ml** [comment]
LGTM! I think we're good as long as we resolve the conflicts and pass the builds.

**Pull request merged**